1301-x86-include-asm-intel_txt.h-constants-and-accessors-.patch

```diff
@@ -0,0 +1,304 @@
++#define TXTCR_ESTS			0x0008
++#define TXTCR_ERRORCODE			0x0030
++#define TXTCR_CMD_RESET			0x0038
++#define TXTCR_CMD_CLOSE_PRIVATE		0x0048
++#define TXTCR_DIDVID			0x0110
++#define TXTCR_VER_EMIF			0x0200
++#define TXTCR_CMD_UNLOCK_MEM_CONFIG	0x0218
++#define TXTCR_SINIT_BASE		0x0270
++#define TXTCR_SINIT_SIZE		0x0278
++#define TXTCR_MLE_JOIN			0x0290
++#define TXTCR_HEAP_BASE			0x0300
++#define TXTCR_HEAP_SIZE			0x0308
++#define TXTCR_SCRATCHPAD		0x0378
++#define TXTCR_CMD_OPEN_LOCALITY1	0x0380
++#define TXTCR_CMD_CLOSE_LOCALITY1	0x0388
++#define TXTCR_CMD_OPEN_LOCALITY2	0x0390
++#define TXTCR_CMD_CLOSE_LOCALITY2	0x0398
++#define TXTCR_CMD_SECRETS		0x08e0
++#define TXTCR_CMD_NO_SECRETS		0x08e8
++#define TXTCR_E2STS			0x08f0
++
++/*
++ * Secure Launch Defined Error Codes used in MLE-initiated TXT resets.
++ *
++ * TXT Specification
++ * Appendix I ACM Error Codes
++ */
++#define SL_ERROR_GENERIC		0xc0008001
++#define SL_ERROR_TPM_INIT		0xc0008002
++#define SL_ERROR_TPM_INVALID_LOG20	0xc0008003
++#define SL_ERROR_TPM_LOGGING_FAILED	0xc0008004
++#define SL_ERROR_REGION_STRADDLE_4GB	0xc0008005
++#define SL_ERROR_TPM_EXTEND		0xc0008006
++#define SL_ERROR_MTRR_INV_VCNT		0xc0008007
++#define SL_ERROR_MTRR_INV_DEF_TYPE	0xc0008008
++#define SL_ERROR_MTRR_INV_BASE		0xc0008009
++#define SL_ERROR_MTRR_INV_MASK		0xc000800a
++#define SL_ERROR_MSR_INV_MISC_EN	0xc000800b
++#define SL_ERROR_INV_AP_INTERRUPT	0xc000800c
++#define SL_ERROR_INTEGER_OVERFLOW	0xc000800d
++#define SL_ERROR_HEAP_WALK		0xc000800e
++#define SL_ERROR_HEAP_MAP		0xc000800f
++#define SL_ERROR_REGION_ABOVE_4GB	0xc0008010
++#define SL_ERROR_HEAP_INVALID_DMAR	0xc0008011
++#define SL_ERROR_HEAP_DMAR_SIZE		0xc0008012
++#define SL_ERROR_HEAP_DMAR_MAP		0xc0008013
++#define SL_ERROR_HI_PMR_BASE		0xc0008014
++#define SL_ERROR_HI_PMR_SIZE		0xc0008015
++#define SL_ERROR_LO_PMR_BASE		0xc0008016
++#define SL_ERROR_LO_PMR_MLE		0xc0008017
++#define SL_ERROR_INITRD_TOO_BIG		0xc0008018
++#define SL_ERROR_HEAP_ZERO_OFFSET	0xc0008019
++#define SL_ERROR_WAKE_BLOCK_TOO_SMALL	0xc000801a
++#define SL_ERROR_MLE_BUFFER_OVERLAP	0xc000801b
++#define SL_ERROR_BUFFER_BEYOND_PMR	0xc000801c
++#define SL_ERROR_OS_SINIT_BAD_VERSION	0xc000801d
++#define SL_ERROR_EVENTLOG_MAP		0xc000801e
++#define SL_ERROR_TPM_NUMBER_ALGS	0xc000801f
++#define SL_ERROR_TPM_UNKNOWN_DIGEST	0xc0008020
++#define SL_ERROR_TPM_INVALID_EVENT	0xc0008021
++
++#define TXT_OS_MLE_MAX_VARIABLE_MTRRS	32
++
++#define SLAUNCH_BOOTLOADER_MAGIC	0x4c534254
++
++#ifndef __ASSEMBLY__
++
++/* We need to differentiate between pre- and post paging enabled. */
++#ifdef __BOOT_DEFS_H__
++#define _txt(x) _p(x)
++#else
++#include <xen/types.h>
++#include <asm/page.h>	// __va()
++#define _txt(x) __va(x)
++#endif
++
++/*
++ * Always use private space as some of registers are either read-only or not
++ * present in public space.
++ */
++static inline volatile uint64_t read_txt_reg(int reg_no)
++{
++	volatile uint64_t *reg = _txt(TXT_PRIV_CONFIG_REGS_BASE + reg_no);
```

Comment:
I think that macro is for accessing symbols and not constructed addresses.  Couldn't find any other suitable one.

Reply:
> Why do you think it's not suitable for constructed addresses?

Not that it can't be used, it just requires a typed lvalue. I can use it like
```c
return ACCESS_ONCE(*(uint64_t *)_txt(TXT_PRIV_CONFIG_REGS_BASE + reg_no));
```
but I can't include `<xen/lib.h>` in the early code:
```
./include/xen/cache.h:4:10: fatal error: asm/cache.h: No such file or directory
```
Would have to extract at least `ACCESS_ONCE` to a separate header.